### PR TITLE
Make sign-windows action more composable+testable

### PR DESCRIPTION
**Pull request by bennettgoble**
This changeset introduces a revised version of `sign-pkg-windows` called `sign-windows`. The new action is more composable than the previous behavior, which makes it easier to test, update and use in different contexts.

## Testing

Testing has been expanded: pytest coverage has been added for python code, and the sign-windows action will be exercised in CI.

## Updated sign-windows action

The sign-windows action no longer makes any assumptions about the shape of the artifacts it needs to sign and package nor does it download artifacts on behalf of the consuming package. This makes the resulting action, `sign-windows` more composable and testable.

Responsibility of building the NSIS installer has been delegated to a fork of [joncloud/makensis-action](https://github.com/joncloud/makensis-action), which saves us from needing to scour the PATH for NSIS.

<details>
  <summary>Before (build.yaml)</summary>

```yaml
  sign-and-package-windows:
    needs: build
    runs-on: windows
    steps:
      - name: Sign and package Windows viewer
        uses: secondlife/viewer-build-util/sign-pkg-windows@main
        with:
          vault_uri: "${{ secrets.AZURE_KEY_VAULT_URI }}"
          cert_name: "${{ secrets.AZURE_CERT_NAME }}"
          client_id: "${{ secrets.AZURE_CLIENT_ID }}"
          client_secret: "${{ secrets.AZURE_CLIENT_SECRET }}"
          tenant_id: "${{ secrets.AZURE_TENANT_ID }}"
```
</details>

<details>
  <summary>After (build.yaml)</summary>

```yaml
  sign-and-package-windows:
    needs: build
    runs-on: windows
    steps:
      - name: Fetch Windows app
        uses: actions/download-artifact@v3
        with:
          name: Windows-app
          path: .app

      - name: Sign installer components
        uses: secondlife/viewer-build-util/sign-windows@v2
        with:
          vault-uri: "${{ secrets.AZURE_KEY_VAULT_URI }}"
          cert-name: "${{ secrets.AZURE_CERT_NAME }}"
          client-id: "${{ secrets.AZURE_CLIENT_ID }}"
          client-secret: "${{ secrets.AZURE_CLIENT_SECRET }}"
          tenant-id: "${{ secrets.AZURE_TENANT_ID }}"
          files-to-sign: |
            .app/SecondLifeViewer.exe
            .app/SLVersionChecker.exe
            .app/llplugin/dullahan_host.exe

      - name: Build installer
        id: nsis
        uses: secondlife-3p/makensis-action@v4
        with:
          script-file: .app/secondlife_viewer_tmp.nsi
          arguments: /V2

      - name: Sign installer
        uses: secondlife/viewer-build-util/sign-windows@v2
        with:
          vault-uri: "${{ secrets.AZURE_KEY_VAULT_URI }}"
          cert-name: "${{ secrets.AZURE_CERT_NAME }}"
          client-id: "${{ secrets.AZURE_CLIENT_ID }}"
          client-secret: "${{ secrets.AZURE_CLIENT_SECRET }}"
          tenant-id: "${{ secrets.AZURE_TENANT_ID }}"
          files-to-sign: "*_Setup.exe"
```
</details>

## Breaking changes

This represents a breaking change to viewer-build-utils. While this is common in github actions, **it appears that we are using the unfortunate `viewer-build-utils@main` tag rather than a major version tag such as `viewer-build-utils@v1` in [secondlife/viewer][]**. 😬 This should be changed.

## Odds and ends:

- Fixed inconsistency in action kebab and snake case arguments

## TODO

- [x] Prepare corresponding branch in [secondlife/viewer][]
- [x] Figure out how to work around problematic `main` branch tag being used in the viewer

[secondlife/viewer]: https://github.com/secondlife/viewer

### CODE REVIEW

File: viewerutil/sign_windows.py [hunk 1]
```diff
@@ -1,42 +1,25 @@
-#!/usr/bin/env python3
-"""\
-@file   sign.py
-@author Nat Goodspeed
-@date   2023-09-14
-@brief  Sign the designated executable using Microsoft AzureSignTool.
-
-$LicenseInfo:firstyear=2023&license=viewerlgpl$
-Copyright (c) 2023, Linden Research, Inc.
-$/LicenseInfo$
```
Comment (bennettgoble) [inline viewerutil/sign_windows.py:10]: These kinds of machine readable license headers [are not needed anymore](https://wiki.secondlife.com/wiki/Coding_standard#MIT_Licensed_Code), a LICENSE in root suffices.

Comment (nat-goodspeed) [reply viewerutil/sign_windows.py:10]: So noted, thanks.

**Comment by nat-goodspeed** [comment]
> Figure out how to work around problematic main branch tag being used in the viewer

I surmise that the correct sequence would be something like:

1. Tag the current main tip as v1.0, also as v1.
2. Update the viewer repo to reference `viewer-build-util@v1`.
3. Tag this branch as v2.0, also as v2.
4. Prepare the consuming viewer branch to reference `viewer-build-util@v2`.
5. Test and merge the consuming viewer branch.

Once the viewer build is referencing `v1`, the timing of merging `viewer-build-util`'s `v2` branch to `main` becomes less critical.

I remember that you mentioned a tool that could automatically generate and update the shorthand `v1`, `v2` etc. tags, but I don't remember details.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      - uses: actions/checkout@v4
+
+      - name: Test sign-windows (Multiple files)
+        uses: ./sign-windows
+        with:
+          vault-uri: "${{ secrets.AZURE_KEY_VAULT_URI }}"
+          cert-name: "${{ secrets.AZURE_CERT_NAME }}"
+          client-id: "${{ secrets.AZURE_CLIENT_ID }}"
+          client-secret: "${{ secrets.AZURE_CLIENT_SECRET }}"
+          tenant-id: "${{ secrets.AZURE_TENANT_ID }}"
```
Comment (nat-goodspeed) [inline .github/workflows/ci.yaml:42]: These secrets are stored for the `secondlife/viewer` and `secondlife/viewer-private` repos, not for `viewer-build-util`. Or did you store them here too?

File: sign-windows/action.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      run: which sign-windows || pip install -e "$(dirname ${{ github.action_path }})/.."
+
+    - name: Install AzureSignTool
+      shell: bash
+      run: which azuresigntool || dotnet tool install --global AzureSignTool
+
+    - name: Sign files
+      shell: bash
+      env:
+        FILES_TO_SIGN: ${{ inputs.files-to-sign }}
```
Comment (nat-goodspeed) [inline sign-windows/action.yaml]: It looks as though this and the `for` loop on line 41 won't correctly handle pathnames containing spaces.

**Comment by bennettgoble** [comment]
> I remember that you mentioned a tool that could automatically generate and update the shorthand `v1`, `v2` etc. tags, but I don't remember details.

You may be pleased to find the details here: https://github.com/secondlife/viewer-build-util/pull/5

File: test/sign_windows_test.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        run.return_value = run_return
+        with self.assertRaises(SignError) as ctx:
+            sign('executable-1',
+                 vault_uri='vault_uri',
+                 cert_name='cert_name',
+                 client_id='client_id',
+                 client_secret='client_secret',
+                 tenant_id='tenant_id',
+                 )
+        self.assertTrue('executable-1 signing failed' in str(ctx.exception))
```
Comment (nat-goodspeed) [inline test/sign_windows_test.py:26]: I like this test; thanks for the `unittest.mock.MagicMock` example.

The tests below exercise functionality we should remove: the new `AzureSignTool` doesn't report cert expiration the way its predecessor `signtool` does. Every run we get the "Failed to find certificate expiration date" warning.

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (bennettgoble) [reply .github/workflows/ci.yaml:42]: I've set this repository up with its own AZURE_ secrets in order to exercise the action in an end-to-end fashion.

File: test/sign_windows_test.py (hunk 4 above)
Comment (bennettgoble) [reply test/sign_windows_test.py:26]: Alright, so should the certificate expiration date warning functionality be removed completely?

File: sign-windows/action.yaml (hunk 3 above)
Comment (bennettgoble) [reply sign-windows/action.yaml]: Ah, you're right. I've moved the multiple-file handling to python to address this and added a test. Thanks!

**Comment by bennettgoble** [comment]
Alright, I've created a new branch, `next` and retargeted this PR. It should be safe to merge breaking changes into this new branch as it leaves main untouched.

**Comment by bennettgoble** [comment]
@nat-goodspeed lmk if the certificate expiration feature should get removed. This can be done in a follow-up PR. I'm going to merge this into the WIP branch `next` which will comprise a v2 (or v3) release.

**Comment by nat-goodspeed** [comment]
Yes: since AzureSignTool doesn't report cert expiration at all, we can remove all the code to both hunt for it and test for it. In fact we need no longer capture AzureSignTool's output.

**Pull request merged**